modules/ROOT/pages/idt.adoc

```diff
@@ -0,0 +1,200 @@
+* *P:* Present bit. Must be set (*1*) for the descriptor to be valid.
+
+For more information, see <<4>>.
+
+== Structure on x86-64
+
+=== Table
+
+On 64-bit processors, the entries in the *IDT* are 16 bytes long and form a table in memory like this:
+
+.*Interrupt Descriptor Table*
+[cols="2,1"]
+|===
+|*Address*
+|*Content*
+
+|IDTR Offset + 0
+|Entry 0
+
+|IDTR Offset + 16
+|Entry 1
+
+|IDTR Offset + 32
+|Entry 2
+
+|...
+|...
+
+|IDTR Offset + 4080
+|Entry 255
+|===
+
+The total size of the table including all 256 entries is 4096 bytes, which is conveniently the size of a single *Page*.
+
+The corresponding entry for a given *Interrupt Vector* is pointed to in memory by scaling the vector by 16 and adding it to the value of the *Offset* field in the *IDTR*.
+
+If the CPU is running in xref:x86.adoc#_long_mode_x86_64_only[*Long Mode*] or an associated xref:x86.html#_compatibility_mode_x86_64_only[*Compatibility Mode*], then the CPU will enter *Long Mode* after calling the *ISR*. To return in this case, the _IRETQ_ instruction should be used.
+
+For more information, see <<5>>.
+
+=== Interrupt Descriptor
+
+Each entry in the table has a complex structure:
+
+*TODO DIAGRAM OR TABLE HERE*
+
+* *Offset:* A 64-bit value, split in three parts. It is the virtual address of the entry point of an *Interrupt Service Routine*.
+* *Selector:* A 16-bit value with multiple fields which must refer to a valid code segment in the *GDT*. For more information, see <<3>>.
+* *Gate Type:* A 4-bit value which defines the type of gate this **Interrupt Descriptor** represents. There are two valid type values:
+** *0xE:* 64-bit Interrupt Gate.
+** *0xF:* 64-bit Trap Gate.
+* *DPL:* A 2-bit value which defines the *CPU Privilege Levels* which are allowed to access this interrupt via the _INT_ instruction. Hardware interrupts ignore this mechanism.
+* *P:* Present bit. Must be set (*1*) for the descriptor to be valid.
+
+For more information, see <<6>>.
+
+== Gate Types
+
+There are at a basic level two kinds of interrupts: undesired interrupts or *Exceptions*, which occur when running code has encountered a CPU-level error, and desired interrupts such as hardware interrupts or the _INT_ instruction. In the first case it is pertinent to save the address of the _currently_ executing instruction so that it can be retried, this is called a *Trap*. In the second case it is pertinent to save the address of the _next_ instruction so that execution can be resumed where it left off. Another difference to note is that with *Traps*, new interrupts might occur during the service routine, but when the CPU is serving an *IRQ* or _INT_ call, further interrupts are masked until an *End of Interrupt* signal is sent. How a certain interrupt is served depends on which kind of gate you put in the corresponding *Interrupt Descriptor*.
```

Comment:
CPU exceptions are not always undesired.

Reply:
Statistically, most exceptions are desired since shit doesn't crash often but it sure does demand page often.